--- Decoder.sdnet.original	2018-05-17 17:11:51.420905795 -0400
+++ Decoder.sdnet	2018-05-17 12:48:49.678883281 -0400
@@ -154,6 +154,31 @@
     method move_to_section = done(0);
     method increment_offset = 0; } }
 class Decoder_0_t::UserEngine(200, 0) {
+  Packet_input packet_in;
+  Packet_output packet_out;
+  class Update_fl_t::Tuple(inout) {
+    struct {
+      k_1 : 8,
+      packet_count_1 : 8 } }
+  class hdr_t_0::Tuple(inout) {
+    struct {
+      eth : eth_h,
+      fec : fec_h } }
+  class ioports_t::Tuple(inout) {
+    struct {
+      ingress_port : 4,
+      egress_port : 4 } }
+  class local_state_t::Tuple(inout) {
+    struct {
+      id : 16 } }
+  class Parser_extracts_t::Tuple(inout) {
+    struct {
+      size : 32 } }
+  Update_fl_t Update_fl;
+  hdr_t_0 hdr;
+  ioports_t ioports;
+  local_state_t local_state;
+  Parser_extracts_t Parser_extracts;
   class Decoder_input_t::Tuple(in) {
     struct {
       stateful_valid : 1,
@@ -251,10 +276,15 @@
     Update_lvl_0.hdr = Update_lvl.hdr,
     Update_lvl_0.ioports = Update_lvl.ioports,
     Update_lvl_0.local_state = Update_lvl.local_state,
-    Update_lvl_1.Update_fl = Update_lvl_0.Update_fl,
-    Update_lvl_1.hdr = Update_lvl_0.hdr,
-    Update_lvl_1.ioports = Update_lvl_0.ioports,
-    Update_lvl_1.local_state = Update_lvl_0.local_state,
+    Decoder_0.Update_fl = Update_lvl_0.Update_fl,
+    Decoder_0.hdr = Update_lvl_0.hdr,
+    Decoder_0.ioports = Update_lvl_0.ioports,
+    Decoder_0.local_state = Update_lvl_0.local_state,
+    Decoder_0.Parser_extracts = Parser.Parser_extracts,
+    Update_lvl_1.Update_fl = Decoder_0.Update_fl,
+    Update_lvl_1.hdr = Decoder_0.hdr,
+    Update_lvl_1.ioports = Decoder_0.ioports,
+    Update_lvl_1.local_state = Decoder_0.local_state,
     Decoder_0.Decoder_input = Update_lvl_0.Decoder_input,
     Update_lvl_1.Decoder_output = Decoder_0.Decoder_output,
     Update_lvl.hdr = Parser.hdr,
@@ -262,6 +292,7 @@
     Update_lvl.ioports = ioports,
     ioports = Update_lvl_1.ioports,
     Parser.packet_in = packet_in,
-    Deparser.packet_in = Parser.packet_out,
+    Decoder_0.packet_in = Parser.packet_out,
+    Deparser.packet_in = Decoder_0.packet_out,
     packet_out = Deparser.packet_out,
-    Deparser.Deparser_extracts = Parser.Parser_extracts } }
\ No newline at end of file
+    Deparser.Deparser_extracts = Decoder_0.Parser_extracts } }
