--- Encoder.sdnet	2019-01-06 09:14:48.204044763 -0500
+++ Encoder.sdnet.original	2019-01-06 09:13:57.155746690 -0500
@@ -207,6 +207,33 @@
     method move_to_section = done(0);
     method increment_offset = 0; } }
 class headerCompress_0_t::UserEngine(1000, 0) {
+  Packet_input packet_in;
+  Packet_output packet_out;
+  class CheckTcp_fl_t::Tuple(inout) {
+    struct {
+      forward_1 : 1 } }
+  class hdr_t_0::Tuple(inout) {
+    struct {
+      eth : eth_h,
+      fec : fec_h,
+      ipv4 : ipv4_t,
+      tcp : tcp_h } }
+  class ioports_t::Tuple(inout) {
+    struct {
+      ingress_port : 4,
+      egress_port : 4 } }
+  class local_state_t::Tuple(inout) {
+    struct {
+      id : 16 } }
+  class Parser_extracts_t::Tuple(inout) {
+    struct {
+      size : 32 } }
+  CheckTcp_fl_t CheckTcp_fl;
+  hdr_t_0 hdr;
+  ioports_t ioports;
+  local_state_t local_state;
+  Parser_extracts_t Parser_extracts;
+
   class headerCompress_input_t::Tuple(in) {
     struct {
       stateful_valid : 1 } }
@@ -369,17 +396,26 @@
     CheckTcp_lvl_0.hdr = CheckTcp_lvl.hdr,
     CheckTcp_lvl_0.ioports = CheckTcp_lvl.ioports,
     CheckTcp_lvl_0.local_state = CheckTcp_lvl.local_state,
-    CheckTcp_lvl_1.CheckTcp_fl = CheckTcp_lvl_0.CheckTcp_fl,
-    CheckTcp_lvl_1.hdr = CheckTcp_lvl_0.hdr,
-    CheckTcp_lvl_1.ioports = CheckTcp_lvl_0.ioports,
-    CheckTcp_lvl_1.local_state = CheckTcp_lvl_0.local_state,
+    
+    CheckTcp_lvl_1.CheckTcp_fl = headerCompress_0.CheckTcp_fl,
+    CheckTcp_lvl_1.hdr = headerCompress_0.hdr,
+    CheckTcp_lvl_1.ioports = headerCompress_0.ioports,
+    CheckTcp_lvl_1.local_state = headerCompress_0.local_state,
+    
     headerCompress_0.headerCompress_input = CheckTcp_lvl_0.headerCompress_input,
+    headerCompress_0.CheckTcp_fl = CheckTcp_lvl_0.CheckTcp_fl,
+    headerCompress_0.hdr = CheckTcp_lvl_0.hdr,
+    headerCompress_0.ioports = CheckTcp_lvl_0.ioports,
+    headerCompress_0.local_state = CheckTcp_lvl_0.local_state,
+
     CheckTcp_lvl_1.headerCompress_output = headerCompress_0.headerCompress_output,
     CheckTcp_lvl.hdr = Parser.hdr,
     Deparser.hdr = CheckTcp_lvl_1.hdr,
     CheckTcp_lvl.ioports = ioports,
     ioports = CheckTcp_lvl_1.ioports,
     Parser.packet_in = packet_in,
-    Deparser.packet_in = Parser.packet_out,
+    headerCompress_0.packet_in = Parser.packet_out,
+    Deparser.packet_in = headerCompress_0.packet_out,
     packet_out = Deparser.packet_out,
-    Deparser.Deparser_extracts = Parser.Parser_extracts } }
\ No newline at end of file
+    headerCompress_0.Parser_extracts = Parser.Parser_extracts, 
+    Deparser.Deparser_extracts = headerCompress_0.Parser_extracts } }
